### DISCUSSION TITLE
What's the proper way to go about `render_to_texture` in 2D?

### DISCUSSION THREAD

**Discussion by Teajey** [bevyengine/bevy#5627] [Q&A] [answered]
I'm in over my head trying to mimic the `3d/render_to_texture.rs` example in 2D so that I can render my game in a fixed-size image inside the window. I'm getting this error:
```
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 1, Vulkan)>`
    In a pass parameter
      note: command buffer = `<CommandBuffer-(0, 1, Vulkan)>`
    Attempted to use a texture (9, 1, Vulkan) mips 0..1 layers 0..1 with conflicting usages. Current usage RESOURCE and new usage COLOR_TARGET. COLOR_TARGET is an exclusive usage and cannot be used with any otherusages within the usage scope (renderpass or compute dispatch).
```
Does anyone have any immediate insight into what this means? I've basically just initialized an `Image` in the exact same way as in `3d/render_to_texture.rs`, but set it as the render target of a `Camera2dBundle` rather than a `Camera3dBundle`. Perhaps `Image` requires a different configuration for 2D? I wouldn't know where to start

**Comment by grace125** [comment]
I got the same error because I didn't add the `RenderLayers` component to the 2d camera that rendered to texture. This seems to also happen whenever this camera's `RenderLayers` component contains layer `0`. [Here's some docs](https://docs.rs/bevy/latest/bevy/render/view/struct.RenderLayers.html) if it's at all helpful.

**Reply by Teajey** [reply]
I'm using `let second_pass_layer = RenderLayers::layer(1);` which I've inserted on the `Image`, and the `Camera2dBundle` with `priority: 1` that renders that `Image`. 

It's flipped from the example `3d/render_to_texture.rs` which instead attaches the extra render layer to the first pass. But I think what I'm doing is right because I want to render everything to the `Image` by default, and then render the `Image` to the window in the second pass.

**Comment by Teajey** [comment, marked as answer]
I've found my mistake was attaching my render target to an `ImageBundle` rather than a `SpriteBundle`. That's about the extent of my understanding. 

Now I'm dealing with a problem where my UI is rendering to both cameras, but that's another story 🙂🙂

It would be really nice to see an [example](https://github.com/bevyengine/bevy/tree/main/examples) of how to properly render to texture in 2D; would probably make for a simpler example too!

**Comment by fossegutten** [comment]
Do you have example code? I am struggling with some similar errors

**Reply by Teajey** [reply]
What does your error look like?

**Comment by rumbogs** [comment]
Hey @Teajey , I have the exact same requirement, can you share an example of how you fixed it?

Not sure I understand what `ImageBundle` you are referring to.

In my case I'm attaching the RenderTarget to the Camera:

```
                    commands.spawn_bundle(Camera2dBundle {
                        camera: Camera {
                            target: RenderTarget::Image(image_handle.clone()),
                            // renders after the main camera which has default value: 0
                            priority: 1,
                            ..default()
                        },
                        ..default()
                    })

```

**Reply by Teajey** [reply]
I'm referring to [`ImageBundle`](https://docs.rs/bevy/0.8.1/bevy/ui/entity/struct.ImageBundle.html) and [`SpriteBundle`](https://docs.rs/bevy/0.8.1/bevy/prelude/struct.SpriteBundle.html). Are you seeing the exact same `wgpu error: Validation Error` as me?

In my case I fixed this error by rendering the image to a `SpriteBundle` instead of an `ImageBundle`:
```diff
    commands
--      .spawn_bundle(ImageBundle {
--          image: UiImage::from(image_handle),
++      .spawn_bundle(SpriteBundle {
++          texture: image_handle,
            ..default()
        })
        .insert(second_pass_layer)
        .insert(GameScreen);
```

**Reply by rumbogs** [reply]
Thanks for the reply. Almost the same, the command buffer is a bit different, but I think that's expected since it's a different arhitecture. I'm rendering to a Mesh 2D though so this doesn't apply to my issue.

**Comment by hakolao** [comment]
I don't think using a `SpriteBundle` really is the solution here. It's a workaround. However, you should be able to use `ImageBundle` by disabling ui rendering from the secondary camera as explained here in #6181.

Add `UiCameraConfig { show_ui: false },` to the camera with `second_pass_layer`.
